PERF: avoid allocating column names where possible

SamSaffron opened the pull request:
When requesting columns names from database adapters AR:Result
would dup/freeze column names, this prefers using fstrings which
cuts down on repeat allocations

Attributes that are retained keep these fstrings around for the long
term

Note, this has the highest impact on "short" result sets, eg: Topic.first where you can void allocating the number of columns * String. 

See also: https://samsaffron.com/archive/2018/02/16/reducing-string-duplication-in-ruby for detailed explanation of uminus

per @rafaelfranca we are fine including uminus in master now. cc @tenderlove @sgrif

SamSaffron commented:
Very interesting, I would expect map! to be a tiny bit faster (and possibly
more risky) but I did not know &:-@ is faster, do you mind micro benching
this? I would love to see the numbers and understand how MRI optimises
there.

On Tue, Jun 5, 2018 at 12:22 PM, Bart <notifications@github.com> wrote:

> *@bdewater* commented on this pull request.
> ------------------------------
>
> In activerecord/lib/active_record/result.rb
> <https://github.com/rails/rails/pull/33051#discussion_r192929740>:
>
> > @@ -125,7 +125,7 @@ def hash_rows
>            begin
>              # We freeze the strings to prevent them getting duped when
>              # used as keys in ActiveRecord::Base's @attributes hash
> -            columns = @columns.map { |c| c.dup.freeze }
> +            columns = @columns.map { |c| -c }
>
> Might want to change to columns = @columns.map(&:-@) for a small speed
> improvement.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/33051#pullrequestreview-125800510>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAUXfm9Rs7lKMmtCTtQLmE3a_S1vOaWks5t5etogaJpZM4UYh32>
> .
>

bdewater commented:
It's the reason Rubocop has Style/SymbolProc. There's a lot of context in https://github.com/rails/rails/pull/16833

`benchmark.rb` for this case:
```ruby
require 'benchmark/ips'

COLUMNS = %w(id foo bar baz qux test yolo)

def block
  COLUMNS.map { |c| -c }
end

def to_proc
  COLUMNS.map(&:-@)
end

Benchmark.ips do |x|
  x.report("block")   { block }
  x.report("to_proc") { to_proc }
  x.compare!
end
```

```
Warming up --------------------------------------
               block    65.000k i/100ms
             to_proc    71.129k i/100ms
Calculating -------------------------------------
               block    774.817k (± 1.4%) i/s -      3.900M in   5.034500s
             to_proc    863.802k (± 1.2%) i/s -      4.339M in   5.023677s

Comparison:
             to_proc:   863801.9 i/s
               block:   774816.7 i/s - 1.11x  slower
```

rafaelfranca commented:
LGTM! Let me know when you update the PR to include the `&:-@` change.

SamSaffron committed:
PERF: avoid allocating column names where possible

When requesting columns names from database adapters AR:Result
would dup/freeze column names, this prefers using fstrings which
cuts down on repeat allocations

Attributes that are retained keep these fstrings around for the long
term

Note, this has the highest impact on "short" result sets, eg: Topic.first where you can void allocating the number of columns * String.

SamSaffron commented:
@rafaelfranca  done! 🎊

rafaelfranca commented:
❤️
